### DISCUSSION TITLE
[Share] Trigger `insertCitation` in the note editor

### DISCUSSION THREAD

**Discussion by inertia42** [windingwind/zotero-actions-tags#346] [Action Scripts]
### Description

To easily insert citations in the note editor, I implemented an action script that simulates clicking the `insertCitation` button.

**The version provided by [windingwind](https://github.com/windingwind) in the comments is better.**

### Event

None

### Operation

Script

### Data

```javascript
/**
 * Trigger `insertCitation` in the note editor
 * @author inertia42
 * @usage 
 * @link https://github.com/windingwind/zotero-actions-tags/discussions/346
 * @see https://github.com/windingwind/zotero-actions-tags/discussions/346
 */

const ButtonTitle = "插入引注" // Here you need to provide the corresponding title in the language you want to use.

const document = require("document");

const container1 = document.querySelector('#BetterNotes-editor-main');
const container2 = document.querySelector('#zotero-note-editor');
let iframe;

if (container1) {
  iframe = container1.querySelector('iframe#editor-view');
} else if (container2) {
  iframe = container2.querySelector('iframe#editor-view');
} else {
	return `Not on the note editor page!`
}

if (iframe) {
  const insertButton = iframe.contentDocument.querySelector('button.toolbar-button[title="' + ButtonTitle + '"]');
  insertButton.click();
}
```

### Anything else

However, this script still has a bug. When triggered, it inserts two `{citation}` tags, but only one of them becomes a citation format. When simulating a click using the same DOM element in the debugger, it works normally. I'm not sure if it's due to the note editor or the action plugin. Since I'm not familiar with JavaScript, and I couldn't spot any issues in the note editor's relevant code, I was wondering if anyone knows the reason for this behavior? Thank you very much!

**Comment by windingwind** [comment]
Fixes to the original action:
1. Only trigger once (need `if (item) return;`, see https://github.com/windingwind/zotero-actions-tags?tab=readme-ov-file#custom-script)
2. Support note editor in library, reader, and note tab.

```javascript
/**
 * Trigger `insertCitation` in the note editor
 * @author inertia42 windingwind
 * @usage
 * @link https://github.com/windingwind/zotero-actions-tags/discussions/346
 * @see https://github.com/windingwind/zotero-actions-tags/discussions/346
 */

// Replace with the corresponding tooltip text of the `Insert Citation` button
// const ButtonTitle = "插入引注";
const ButtonTitle = "Insert Citation";

// Skip if in items loop
if (item) return;

const document = require("document");
const ZoteroPane = require("ZoteroPane");
const ZoteroContextPane = require("ZoteroContextPane");
const Zotero_Tabs = require("Zotero_Tabs");

let editor;
switch (Zotero_Tabs.selectedType) {
  case "reader":
    editor = ZoteroContextPane.activeEditor;
    break;
  case "library":
    editor = ZoteroPane.itemPane._noteEditor;
    break;
  case "note":
    editor = Zotero.BetterNotes?.api.workspace
      .getWorkspaceByTabID()
      ?.querySelector("#BetterNotes-editor-main");
    break;
}

if (!editor) {
  return `Not in the note editor!`;
}

const insertButton = editor._iframe.contentDocument.querySelector(
  'button.toolbar-button[title="' + ButtonTitle + '"]',
);
insertButton?.click();
```

**Reply by inertia42** [reply]
It works! Thank you so much!

**Reply by windingwind** [reply]
:)

FYR, Better Notes plugin offers various useful APIs, granting full control over note content: https://github.com/windingwind/zotero-better-notes?tab=readme-ov-file#-api

**Comment by ulahcherubim** [comment]
For anyone interested, it seems like Zotero is implementing a keyboard shortcut for inserting citation in note editor soon. Here is the commit from yesterday.
https://github.com/zotero/note-editor/commit/4b088619334f4a3ed39d9010115580fee66b6681
